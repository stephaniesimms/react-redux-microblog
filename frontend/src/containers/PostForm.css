.PostForm { 
  border: 2px solid rgba(0,0,0,.125); 
  border-radius: 10px; 
  padding: 10px; 
  margin: auto;
  max-width: 700px;
}