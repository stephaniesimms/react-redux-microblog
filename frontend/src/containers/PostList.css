.PostList {
  margin: auto;
  max-width: 90%;
  align-items: center;
}

.PostList-card {
  display: inline-block;
  width: 40vw;
  max-width: 50%;
  margin-right: 1em;
}

.PostList-card-title {
  padding: 1em 0em 1em 0em;
}