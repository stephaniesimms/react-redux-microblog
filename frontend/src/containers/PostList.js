import React, { Component } from 'react';
// import PostTitle from '../components/PostTitle';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Card from 'react-bootstrap/Card';
import { getTitlesFromAPI, sendVoteToAPI } from '../actions';

// FIXME: need to rerender PostTitle and sendVoteToAPI from there when user clicks vote
// add votes and display posts by popularity

/** Show list of blog titles, ordered by popularity. */
class PostList extends Component {

  async componentDidMount() {
    if (this.props.titles.length === 0) {
      await this.props.getTitlesFromAPI();
    }
  }
  
  vote(direction, id) {
    console.log("sending vote to API", id, direction)
    this.props.sendVoteToAPI(id, direction);
  }

  render() {
    if (!this.props.titles) return <b>Loading...</b>;
    console.log(this.props)

    // const titles = this.props.titles.map(
    //   (title) =>
    //     <PostTitle key={title.id}
    //       id={title.id}
    //       title={title.title}
    //       description={title.description}
    //       votes={title.votes}
    //       doVote={this.vote} />
    // );

    const titleList = this.props.titles.map(
      (title) =>
        <Card key={title.id} id={title.id}
          style={{ width: '40vw', display: 'inline-block', marginRight: '15px', maxWidth: '300px' }}>
          <Link to={`/posts/${title.id}`}>
            <Card.Title style={{ marginTop: '15px' }}>{title.title}</Card.Title>
          </Link>

          <Card.Body>{title.description}</Card.Body>
          <div className="card-footer">
            <small>{title.votes} votes</small>
            <i className="fas fa-thumbs-up text-success ml-2"
              onClick={() => this.vote("up", title.id)} />
            <i className="fas fa-thumbs-down text-danger ml-2"
              onClick={() => this.vote("down", title.id)} />
          </div>
        </Card>

    );

    return (
      <div>{titleList}</div>
    );
  }
}

function mapStateToProps(state) {
  return { titles: state.titles };
}

export default connect(
  mapStateToProps,
  { getTitlesFromAPI, sendVoteToAPI }
)(PostList);