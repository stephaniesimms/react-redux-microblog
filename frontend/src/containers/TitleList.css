.TitleList-card {
  margin: 1em 0 1em 0;
}

.TitleList-card-title {
  padding: 1em 0 1em 0;
}