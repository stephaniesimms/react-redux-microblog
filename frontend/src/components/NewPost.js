import React, { Component } from 'react';
import PostForm from '../containers/PostForm';

class NewPost extends Component {

  render() {
    return (
      <div>
        <PostForm formType="New Post" />
      </div>
    );
  }
}

export default NewPost;