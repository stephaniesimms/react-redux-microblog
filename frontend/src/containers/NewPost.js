import React, { Component } from 'react';
import { connect } from 'react-redux';
import PostForm from '../components/PostForm';
import {
  sendPostToAPI,
} from '../actions/actions';

//NOTE: must export connected components for testing
export class NewPost extends Component {
  constructor(props) {
    super(props);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.cancel = this.cancel.bind(this);
  }

  /** Handle new post: adds to backend. */

  handleSubmit({ title, description, body }) {
    this.props.sendPostToAPI({ title, description, body });
    this.props.history.push('/');
  }

  /** Cancel (redirect) */

  cancel() {
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <h3>New Post</h3>
        <PostForm
          save={this.handleSubmit}
        />
      </div>
    );
  }
}


export default connect(
  null, { sendPostToAPI }
)(NewPost);