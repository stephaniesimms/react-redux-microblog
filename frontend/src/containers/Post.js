import React, { Component } from 'react';
import { connect } from 'react-redux';
import PostForm from '../components/PostForm';
import PostDisplay from '../components/PostDisplay';
import CommentList from '../components/CommentList';
import CommentForm from '../components/CommentForm';
import NotFound from '../components/NotFound';
import {
  sendDeleteToAPI,
  sendUpdateToAPI,
  getPostFromAPI,
  sendVoteToAPI,
  sendCommentToAPI,
  deleteCommentFromAPI
} from '../actions/actions';

/** Post:
 *
 * - get post data from API, if not present
 * - allows post to be edited (showEditForm is local state for this)
 * - handles edit form submission
 * - handles add-comment form submission
 * - handles comment-deletion
 * - handles post-deletion
 */

//NOTE: must export connected components for testing
export class Post extends Component {
  constructor(props) {
    super(props);
    this.state = {
      editing: false
    }

    this.handleDelete = this.handleDelete.bind(this);
    this.handleEdit = this.handleEdit.bind(this);
    this.showEditForm = this.showEditForm.bind(this);
    this.addComment = this.addComment.bind(this);
    this.deleteComment = this.deleteComment.bind(this);
    this.vote = this.vote.bind(this);
  }

  /** If we don't have the post, request it from API. */

  async componentDidMount() {
    if (!this.props.posts) {
      await this.props.getPostFromAPI(this.props.id);
    }
  }

  /** Toggle state of editing true/false to show or hide post edit form */
  showEditForm() {
    this.setState( state => ({ editing: !state.editing }));
  }

  /** Handle post deletion: deletes from backend. */

  async handleDelete() {
    const postId = this.props.match.params.postId;
    await this.props.sendDeleteToAPI(postId);

    this.props.history.push('/');
  }

  /** Handle post editing: adds to backend. */

  handleEdit(updatedPost) {
    const postId = this.props.match.params.postId;
    this.props.sendUpdateToAPI(updatedPost, postId);

    this.showEditForm();
  }

  /** Handle adding a comment: adds to backend. */

  addComment(text) {
    this.props.sendCommentToAPI(this.props.post.id, text);
  }

  /** Handle deleting a comment in backend. */

  deleteComment(commentId) {
    this.props.deleteCommentFromAPI(this.props.post.id, commentId);
  }

  /** Handle voting in backend. */

  vote(direction) {
    this.props.sendVoteToAPI(this.props.post.id, direction)
  }

  /** Render:
   *
   * - if backend does not find post, display 404 not found msg
   * - if no post yet, show a loading message
   * - if editing, show the edit form & comments
   * - if not editing, show the post & comments
   */

  render() {
    if (this.props.error) {
      return <NotFound />;
    }

    const post = this.props.post;

    if (!post) {
      return <b>'Loading...'</b>;
    }

    return (
      <div className='container-fluid'>
  
        {this.state.editing 
          ? <PostForm post={post} 
            id={post.id} 
            save={this.handleEdit}
            />
          : <PostDisplay post={post}
          showEditForm={this.showEditForm}
          deletePost={this.handleDelete}
          doVote={this.vote} />}

        <section c>
          <hr></hr>
          <h4 className="mb-3">Comments</h4>
          <CommentList comments={post.comments}
            deleteComment={this.deleteComment}
            postId={post.id}
          />
          <CommentForm submitCommentForm={this.addComment} />
        </section>
      </div>
    );
  }
}

function mapDispatchToProps(state, props) {
  let id = props.match.params.postId;
  return {
    id,
    post: state.posts[id],
    error: state.error
  };
}

export default connect(
  mapDispatchToProps,
  {
    sendDeleteToAPI,
    sendUpdateToAPI,
    getPostFromAPI,
    sendCommentToAPI,
    deleteCommentFromAPI,
    sendVoteToAPI,
  }
)(Post);