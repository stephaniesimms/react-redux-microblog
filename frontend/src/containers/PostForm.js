import React, { Component } from 'react';
import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col'
import './PostForm.css';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { sendPostToAPI } from '../actions/actions';

// TODO: should PostForm be a component instead? 
// maybe refactor to follow solution
// otherwise update docstring

/** Show post form.
 *
 * Can be used for editing/adding --- it just calls the parent with edit
 * or cancel actions.
 *
 */
class PostForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
      title: '',
      description: '',
      body: '',
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentDidMount() {
    if (this.props.post) {
      const { title, description, body } = this.props.post;

      this.setState({
        title,
        description,
        body,
      });
    }
  }

  handleChange(evt) {
    this.setState({
      [evt.target.name]: evt.target.value
    });
  };

  /** Checks the formType prop sent from Post component to decide 
   * whether to create a new post or edit an existing post
  */
  async handleSubmit(evt) {
    evt.preventDefault();

    if (this.props.formType === 'Edit') {
      this.props.updatePost({
        ...this.state,
        comments: this.props.post.comments
      });
    } else {
      const { title, description, body } = this.state;
      await this.props.sendPostToAPI({ title, description, body });

      this.props.history.push('/');
    }
  };

  render() {
    return (
      <div className='PostForm'>
        <h3>{this.props.formType}</h3>
        <Form onSubmit={this.handleSubmit}>
          <Form.Group as={Row}>
            <Form.Label column sm={2}>Title</Form.Label>
            <Col sm={10}>
              <Form.Control id='title'
                name='title'
                onChange={this.handleChange}
                value={this.state.title}
                size='md'
                type='text'
                alt='title' />
            </Col>
          </Form.Group>
          <Form.Group as={Row}>
            <Form.Label column sm={2}>Description</Form.Label>
            <Col sm={10}>
              <Form.Control id='description'
                name='description'
                onChange={this.handleChange}
                value={this.state.description}
                size='md'
                type='text'
                alt='description' />
            </Col>
          </Form.Group>
          <Form.Group as={Row}>
            <Form.Label column sm={2}>Body</Form.Label>
            <Col sm={10}>
              <Form.Control id='body'
                name='body'
                onChange={this.handleChange}
                value={this.state.body}
                size='md'
                as='textarea'
                alt='body' rows='5' />
            </Col>
          </Form.Group>
          <div className='row justify-content-end'>
          <Button className='mr-2' type='submit' variant='dark'>
            Post
          </Button>
          <Link to='/'>
              <Button className='mr-3' variant='danger'>
              Cancel
            </Button>
          </Link>
          </div>
        </Form>
      </div>
    );
  }
}

function mapStateToProps(state) {
  return { posts: state.posts };
}


export default connect(mapStateToProps, { sendPostToAPI })(PostForm);