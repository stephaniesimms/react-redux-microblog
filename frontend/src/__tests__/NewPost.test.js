import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { NewPost } from '../containers/NewPost';

it('renders without crashing', function () {
  shallow(<NewPost />);
});

it('matches snapshot', function () {
  let wrapper = shallow(<NewPost />);
  let serialized = toJson(wrapper);
  expect(serialized).toMatchSnapshot();
});