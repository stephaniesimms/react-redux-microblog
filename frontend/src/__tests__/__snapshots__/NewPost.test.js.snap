// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <h3>
    New Post
  </h3>
  <PostForm
    post={
      Object {
        "body": "",
        "description": "",
        "title": "",
      }
    }
    save={[Function]}
  />
</div>
`;
